core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -2474,25 +2474,36 @@ private[spark] class CallerContext(
    val context = "SPARK_" + from + appIdStr + appAttemptIdStr +
      jobIdStr + stageIdStr + stageAttemptIdStr + taskIdStr + taskAttemptNumberStr
 
+   private var callerContextSupported: Boolean = true
+
   /**
    * Set up the caller context [[context]] by invoking Hadoop CallerContext API of
    * [[org.apache.hadoop.ipc.CallerContext]], which was added in hadoop 2.8.
    */
   def setCurrentContext(): Boolean = {
-    var succeed = false
-    try {
-      // scalastyle:off classforname
-      val callerContext = Class.forName("org.apache.hadoop.ipc.CallerContext")
-      val Builder = Class.forName("org.apache.hadoop.ipc.CallerContext$Builder")
-      // scalastyle:on classforname
-      val builderInst = Builder.getConstructor(classOf[String]).newInstance(context)
-      val hdfsContext = Builder.getMethod("build").invoke(builderInst)
-      callerContext.getMethod("setCurrent", callerContext).invoke(null, hdfsContext)
-      succeed = true
-    } catch {
-      case NonFatal(e) => logInfo("Fail to set Spark caller context", e)
+    if (!callerContextSupported) {
+      false
+    } else {
+      try {
+        // scalastyle:off classforname
+        val callerContext = Class.forName("org.apache.hadoop.ipc.CallerContext")
+        val builder = Class.forName("org.apache.hadoop.ipc.CallerContext$Builder")
+        // scalastyle:on classforname
+        val builderInst = builder.getConstructor(classOf[String]).newInstance(context)
+        val hdfsContext = builder.getMethod("build").invoke(builderInst)
+        callerContext.getMethod("setCurrent", callerContext).invoke(null, hdfsContext)
+        true
+      } catch {
+        case e: ClassNotFoundException =>
+          logInfo(
+            s"Fail to set Spark caller context: requires Hadoop 2.8 or later: ${e.getMessage}")
+          callerContextSupported = false
+          false
+        case NonFatal(e) =>
+          logWarning("Fail to set Spark caller context", e)
```

Comment:
Emm, I prefer not to suppress the non-ClassNotFound errors, because they are real errors, compared to the ClassNotFound error, which is rather a *conditional feature* based on the hadoop environment.